refactor (#322)
KxSystems/docs: docs/kb/callbacks.md

If you need to call a remote function that has more than two data parameters, you cannot use implicit parameters on the server as above. You can either define explicit parameters or encapsulate the arguments in a list. We show the latter here.

Define the following on the server,

```q
q)\p 5000                                           / listen on port 5000
q)add3:{x+y+z}                                      / server function
q)proc3:{ echo r:add3 . x; (neg .z.w) (y; r)}       / function for client to call
```

Here the data for `proc3` is passed as a list in the implicit parameter `x` , while the callback function name is passed as `y`. 
Note the use of [`.` (Apply)](../ref/apply.md) to evaluate a non-unary function on a list of arguments.

Now execute the following on the client.

```q
q)clientFunc:{0N!x;}
q)h:hopen `::5000
q)(neg h) (`proc3; 1 2 3; `clientFunc)
  ...
q)hclose h
```

This expression makes an async call to the remote function `proc3`, passing it the list argument `1 2 3` and the symbol `` `clientFunc`` representing the name of the callback function.

The result is that 6 is displayed on the server and then 6 is displayed on the client.

### Function wrapper

> An arbitrary function on the server will not have the appropriate signature to accept a callback. We show here a simple wrapper function that permits any reasonable multivalent function to be called asynchronously with its result returned to the caller.

Reviewer: ```suggestion
An arbitrary function on the server does not have the appropriate signature to accept a callback. This example shows a simple wrapper function that permits any reasonable multivalent function to be called asynchronously with its result returned to the caller.
```

q)(neg h) (`marshal; `add3; 1 2 3; `clientFunc)
  ...
q)hclose h
```

This expression makes an async call to the remote function `marshal`, asking it to invoke the remote function `add3` with list argument `1 2 3` and to pass the result back to `clientFunc`.

The result is that the list is summed on the server and then 6 is displayed on the client.

### Anonymous functions

It is also possible to send a function to be executed remotely on the server using an alternative form of IPC. 
In this case, nothing needs to be defined on the server in advance. Here we show an example sending an anonymous function that returns its value to the client. 

Start a kdb+ server listening on a choosen port e.g.

```bash
$ q -p 5000
```

Execute the following on the client.

```q
q)clientFunc:{0N!x;} 
q)h:hopen `::5000
q)(neg h) ({(neg .z.w) (z; x*y)}; 6; 7; `clientFunc)
```

This expression makes an async call sending: 

- a function that multiplies two arguments and returns the result with a callback
- the arguments 6 and 7
- and the name of `clientFunc` for the callback

The result is that 6 and 7 are multiplied on the server and then 42 is displayed on the client.

!!! warning

>     Give careful consideration before allowing this style IPC in a production environment as a client can bring down an unprotected server.

Reviewer: This warning is lacking. What is the workaround? Can anyone do anything if it brings down the server? This leaves the reader with a dilemma and not much information on how to decide or resolve the problem if it happens.

Author: improved in subsequent commit
